### PR TITLE
[ENHANCEMENT] Prevent using dashboard datasources before they're saved

### PR DESCRIPTION
**Pull request by sinkingpoint**
# Description

Previously, when a dashboard level datasource is added to a dashboard, it is immediately available for use in the dashboard. This causes an error because the server side (e.g. the proxy endpoints) isn't created until the dashboard is saved.

Here we introduce the idea of "saved" datasources, which are the datasources that are available for use in the dashboard because they have been created through a dashboard save.

We then use this list of saved datasources to disable entries in the datasource select dropdown, with an indication that the dashboard needs to be saved before the datasource can be used.

# Screenshots

image

https://github.com/perses/perses/assets/4386405/63a71f53-1636-41fb-8e46-e036bca62643

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [x] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [x] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [ ] Visual tests are stable and unlikely to be flaky. See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests) and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

### CODE REVIEW

**Review by Gladorme — APPROVED** [review]
LGTM, really cool!

**Review by AntoineThebaud — CHANGES_REQUESTED** [review]
Two issues in the current state:
1. You can still encounter the error by changing a saved datasource with mode=direct to mode=proxy (and not saving yet)
2. You cannot use a not-yet-saved datasource that has mode=direct, while the querying would work.

IMHO it's ok to let point 1 aside for now, what you did is already a good improvement. However I would like to be able to use new datasources right away when they have mode=direct.

**Comment by sinkingpoint** [comment]
Updated this with a change to the `EditDatasourcesButton` - we now allow datasources that are direct to be used immediately, and we also _disallow_ datasources that are proxied, but weren't before (i.e. they've been updated to be a proxy). This does introduce a rather gnarly conditional though

File: ui/dashboards/src/components/Datasources/EditDatasourcesButton.tsx [hunk 1]
```diff
@@ -35,6 +36,40 @@ export function EditDatasourcesButton() {
...
+      .filter((key) => {
+        // Datasources are allowed to be used if a) they are direct, or b) they are proxied, and their
+        // proxy is the same as what we have saved.
+        const isDirect = 'directUrl' in (datasources[key]?.plugin?.spec ?? {});
+        const isSavedProxy =
+          !isDirect &&
+          !('directUrl' in (savedDatasources[key]?.plugin?.spec ?? {})) &&
+          datasources[key]?.plugin?.spec?.proxy === savedDatasources[key]?.plugin?.spec?.proxy;
+
+        console.log('name', key, 'isDirect', isDirect, 'isSavedProxy', isSavedProxy);
```
Comment (AntoineThebaud) [inline ui/dashboards/src/components/Datasources/EditDatasourcesButton.tsx:51]: leftover log

**Review by AntoineThebaud — APPROVED** [review]
Awesome, thanks for the fix! LGTM (there's just a remaining console.log to cleanup ;)

**Comment by sinkingpoint** [comment]
Knew I missed something. Fixed!

**Pull request merged**